src/backend/utils/activity/memtrack.c

```diff
@@ -0,0 +1,185 @@
+
+/*
+ * Max backend memory allocation allowed (MB). 0 = disabled.
+ * Max backend bytes is the same but in bytes.
+ * These default to "0", meaning don't check bounds for total memory.
+ */
+int			max_total_memory_mb = 0;
+int64       max_total_memory_bytes = 0;
+
+/*
+ * Private variables for tracking memory use.
+ * These values are preset so memory tracking is active on startup.
+ * After a fork(), they must be reset using 'init_backend_memory()'.
+ */
+PgBackendMemoryStatus  my_memory = INIT_BACKEND_MEMORY;
+PgBackendMemoryStatus  reported_memory = NO_BACKEND_MEMORY;
+int64      allocation_lower_bound = 0;
+int64      allocation_upper_bound = 0;
+
+/* ---------
+ * init_backend_memory() -
+ *
+ * Called immediately after a fork().
+ * Resets private memory counters to their initial startup values
+ */
+void
+init_backend_memory(void)
+{
+	/* Start with nothing allocated. */
+	my_memory = INIT_BACKEND_MEMORY;
+	reported_memory = NO_BACKEND_MEMORY;
+
+	/* Force the next allocation to do global bounds checking. */
+	allocation_lower_bound = 0;
+	allocation_upper_bound = 0;
+}
+
+/*
+ * Clean up memory counters as backend is exiting.
+ *
+ * DSM memory is not automatically returned, so it persists in the counters.
+ * All other memory will disappear, so those counters are set to zero.
+ *
+ * Ideally, this function would be called last, but in practice there are some
+ * late memory releases that happen after it is called.
+ */
+void
+exit_backend_memory(void)
+{
+	/*
+	 * Release non-dsm memory.
+	 * We don't release dsm shared memory since it survives process exit.
+	 */
+	for (int type = 0; type < PG_ALLOC_TYPE_MAX; type++)
+		if (type != PG_ALLOC_DSM)
+			release_backend_memory(my_memory.allocated_bytes_by_type[type], type);
+
+	/* Force the final values to be posted to shmem */
+	update_global_allocation(0, PG_ALLOC_INIT);
+
+	/* If we get a late request, send it to the long path. */
+	allocation_lower_bound = 0;
+	allocation_upper_bound = 0;
+}
+
+
+/*
+ * Update backend memory allocation for a new request.
+ *
+ * There are two versions of this function. This one, which updates
+ * global values in shared memory, and an optimized update_local_allocation()
+ * which only updates private values.
+ *
+ * This version is the "slow path". We invoke it periodically to update
+ * global values and pgstat statistics.
+ *
+ * Note the use of pg_atomic_fetch_add_u64() to update counters
+ * with a signed value assumes we are on a 2's complement machine.
+ * Otherwise, we would need to introduce atomic s64 operators.
+ */
+bool update_global_allocation(int64 size, pg_allocator_type type)
+{
+	int64 delta;
+	int64 dsm_delta;
+
+	/* If we are still initializing, only update the private counters */
+	if (ProcGlobal == NULL || MyProcPid == 0)
+		return update_local_allocation(size, type);
+
+	/* Quick check to verify totals are not negative. Should never happen. */
```

Comment:
Should say "Assert to ...", doesn't matter if it's quick or not, it won't be in a production build.

Reply:
I was thinking of dropping the preface completely and saying "Verify totals are not negative. Should never happen.".

Reply:
Sure, that's fine too.